torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1115,57 +1162,71 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
   std::lock_guard<std::mutex> lock(mutex_);
   abortCommsFromMap(devNCCLCommMap_, abortReason);
   abortCommsFromMap(inInitializationCommMap_, abortReason);
+  return true;
 }
 
 void ProcessGroupNCCL::shutdown() {
   // Don't join threads here since the purpose of this method is to abort all
   // communicators and signal the threads to exit. Joining on the threads could
   // potentially block and hence avoid it in this method.
   terminateProcessGroup_.store(true);
+  workMetaListCV_.notify_one();
 
   std::string abortReason = c10::str("Process Group shutdown on rank ", rank_);
-  abort(abortReason);
+  // lauch abort asynchrounously and wait for it to complete or timeout
+  LOG(INFO) << logPrefix()
+            << "Launching ProcessGroupNCCL abort asynchrounously.";
+  std::future<bool> fut = std::async(std::launch::async, [this, abortReason]() {
+    return this->abort(abortReason);
+  });
 
-  workMetaListCV_.notify_one();
+  waitForFutureOrTimeout(
+      fut, kProcessGroupNCCLAbortTimeout, "ProcessGroup abort");
+  LOG(INFO) << logPrefix() << "ProcessGroupNCCL aborts successfully.";
+
+  // We need to wait for abort to finish before we can safely shut down
+  // heartbeat monitoring thread.
   terminateHeartbeatMonitorThread_.store(true);
   monitorWakeUpCV_.notify_one();
 }
 
 ProcessGroupNCCL::~ProcessGroupNCCL() {
   LOG(INFO) << logPrefix() << "ProcessGroupNCCL destructor entered.";
-  terminateProcessGroup_.store(true);
-  workMetaListCV_.notify_one();
 
+  if (!terminateProcessGroup_.load()) {
```

Comment:
mental note- do you have tests for all 4 combinations of these 2 if's

Reply:
Both true, covered by new test cases
Both false, covered by existing cases
We got covered 3 cases
terminateProcessGroup_ true, abortInDestroyProcessGroup_ false case, we can add a new test case to trigger the destructor
terminateProcessGroup_ false, but abortInDestroyProcessGroup_ is true , but we can add one to print the log